contrib/dynamic_embedding/src/torchrec_dynamic_embedding/dataloader.py

```diff
@@ -0,0 +1,223 @@
+import queue
+import threading
+from typing import Dict, List, Union
+
+import torch
+from torch.utils.data._utils import MP_STATUS_CHECK_INTERVAL
+from torchrec import EmbeddingBagConfig, EmbeddingConfig
+from torchrec.distributed.model_parallel import DistributedModelParallel
+from torchrec.sparse.jagged_tensor import KeyedJaggedTensor
+
+from .id_transformer_group import IDTransformerGroup
+
+
+__all__ = ["wrap", "save"]
+
+
+# Similar to torch.utils.data._utils.pin_memory._pin_memory_loop
+def transform_loop(dataloader, transform_fn, out_queue, done_event):
+    # This setting is thread local, and prevents the copy in pin_memory from
+    # consuming all CPU cores.
+    torch.set_num_threads(1)
+
+    for data in dataloader:
+        if done_event.is_set():
+            break
+        transformed_data = transform_fn(data)
+
+        while not done_event.is_set():
+            try:
+                out_queue.put(transformed_data, timeout=MP_STATUS_CHECK_INTERVAL)
+                break
+            except queue.Full:
+                continue
+        # save memory
+        del transformed_data
+
+
+class DataLoaderIter:
+    def __init__(self, dataloader, transform_fn, num_prefetch=0):
+        self._data_queue = queue.Queue(maxsize=num_prefetch)
+        self._done_event = threading.Event()
+        self._transform_thread = threading.Thread(
+            target=transform_loop,
+            args=(dataloader, transform_fn, self._data_queue, self._done_event),
+        )
+        self._transform_thread.start()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self._get_data()
+
+    def __del__(self):
+        self._done_event.set()
+
+    def _get_data(self):
+        if not self._transform_thread.is_alive():
+            raise RuntimeError("Transform thread exited unexpectedly")
+        data, handles = self._data_queue.get()
+        for handle in handles:
+            handle.wait()
+        return data
+
+
+class DataLoader:
```

Comment:
is this a dataloader? https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader

see https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset

Reply:
This is actually a wrapper of the dataloader. And we could turn this into a `TranPipeline`.